File: Tema 06 - React/petshelter.ui/src/Services/FundraiserService.ts

import axios from "axios";
import { FundaiserSummaryInfo } from "../Models/FundraiserSummaryinfo";
import { FundraiserInformation } from "../Models/FundraiserInformation";
import type { Fundraiser } from "../Models/Fundraiser";

export class FundraiserService {
    private apiUrl: string = "https://localhost:7075";


    public async getAll(): Promise<FundaiserSummaryInfo[]> {
        let fundraisers: any[] = [];
        return await
            axios.get(this.apiUrl + '/Fundraisers')
                .then(response => {
                    fundraisers = response.data;
                    const mappedFundraisers = fundraisers.map(fundraiser => {
                        return new FundaiserSummaryInfo(fundraiser.id, fundraiser.name, fundraiser.status);
                    });
                    return mappedFundraisers;
                })
    }


    public getFundraiserById(id: number): Promise<FundraiserInformation> {
        return axios
        .get(this.apiUrl + '/Fundraisers/' + id)
        .then(response => {
            console.log(response.data);
            var fundraiserResponse: FundraiserInformation = new FundraiserInformation(
                response.data.name,
                response.data.endDate,
                response.data.target,
                response.data.status,
                response.data.amountDonated,
                response.data.donors

            )

            return fundraiserResponse;
        })
    }



    public CreateFundraiser(fundraiser: FundraiserDto): Promise<void> {
            
            let createdFundraiser: FundraiserDto = {
                name: fundraiser.name,
                description: fundraiser.description,
                endDate : fundraiser.endDate,
                target : fundraiser.target,
                owner : {
                    dateOfBirth : new Date(),
                    name: fundraiser.owner.name,
                    idNumber : fundraiser.owner.idNumber
                }
            }
    
            return axios.post(this.apiUrl + '/Fundraisers', createdFundraiser);
    }
    
    public DonateToFundraiser(id: number, amount: number, donor: PersonDto): Promise<void> {

        return axios.post(this.apiUrl + '/Fundraisers/' + id + '/donate?amount='+amount,donor);
    }

    public CloseFundraiser(id: number): Promise<void> {

        return axios.post(this.apiUrl + '/Fundraisers/' + id + '/close');
    }
}


interface PersonDto 
{
    name: string; 
    dateOfBirth: Date | undefined; 
    idNumber: string; 
}

interface FundraiserDto{
    name: string;
    description: string;
    target: number;
    endDate: Date;
    owner: PersonDto;
}

interface FundraiserInformationAllDto{
    Id: number;
    name: string;
    status: string;
}

File: Tema 03- Web API/PetShelter.Domain/Services/FundraiserService.cs

using PetShelter.DataAccessLayer.Models;
using PetShelter.DataAccessLayer.Repository;
using PetShelter.Domain.Extensions.DataAccess;
using PetShelter.Domain.Extensions.DomainModel;
using System.Collections.Immutable;

namespace PetShelter.Domain.Services
{
    public class FundraiserService : IFundraiserService
    {
        private readonly IPersonRepository _personRepository;
        private readonly IDonationRepository _donationRepository;
        private readonly IFundraiserRepository _fundraiserRepository;

        public FundraiserService(IPersonRepository personRepository, IDonationRepository donationRepository, IFundraiserRepository fundraiserRepository)
        {
            _personRepository = personRepository;
            _donationRepository = donationRepository;
            _fundraiserRepository = fundraiserRepository;
        }

        public async Task<int> CreateFundraiser(Fundraiser fundraiser, Person Owner)
        {
            var person = await _personRepository.GetOrAddPersonAsync(Owner.FromDomainModel());
            var fundraisernew = new DataAccessLayer.Models.Fundraiser
            {
                Name = fundraiser.Name,
                Description = fundraiser.Description,
                Owner = person,
                OwnerId = person.Id,
                EndDate = fundraiser.EndDate,
                CreationDate = DateTime.Now,
                Current_Donation_Amount = 0,
                Donations = new List<Donation>(),
                Status = "Active",
                Target = fundraiser.Target
            };
            if (DateTime.Now > fundraiser.EndDate)
                fundraisernew.Status = "Ended";
            await _fundraiserRepository.Add(fundraisernew);
            return fundraisernew.Id;
        }

        public async Task<IReadOnlyCollection<FundraiserSummaryInformation>> GetFundraisersAsync()
        {
            var fundraisers = await _fundraiserRepository.GetAll();
            return fundraisers.Select(p => p.ToFundraiserInfoAll())
                .ToImmutableArray();
        }

        public async Task<FundraiserInformation> GetFundraiser(int id)
        {
            var fundraiser = await _fundraiserRepository.GetById(id);
            if (fundraiser == null)
                return null;
            fundraiser.Owner = await _personRepository.GetById(fundraiser.OwnerId);
            fundraiser.Donations = await _donationRepository.GetDonationsByFundraiserId(id);
            var data = fundraiser.ToFundraiserInformation();
            var donorsids = fundraiser.Donations.Select(d => d.DonorId).Distinct();
            var donors = await _personRepository.GetPersonsByIds(donorsids);
            data.Donors = donors.Select(d => d.ToDomainModel()).ToImmutableArray();
            return data;
        }

        public async Task DonateToFundraiser(int id, Person donor, int amount)
        {
            var fundraiser = await _fundraiserRepository.GetById(id);
            var person = await _personRepository.GetOrAddPersonAsync(donor.FromDomainModel());
            var donation = new Donation
            {
                Amount = amount,
                DonorId = person.Id,
                FundraiserId = fundraiser.Id
            };

            fundraiser.Current_Donation_Amount += amount;

            if (fundraiser.Current_Donation_Amount >= fundraiser.Target)
            {
                fundraiser.Status = "Ended";
            }
            await _donationRepository.Add(donation);
            await _fundraiserRepository.Update(fundraiser);
        }

        public async Task CloseFundraiser(int id)
        {
            var fundraiser = await _fundraiserRepository.GetById(id);
            fundraiser.Status = "Ended";
            await _fundraiserRepository.Update(fundraiser);
        }
    }
}
